### PR TITLE
Add method to retrieve federated session info

### PR DESCRIPTION
**Pull request by ThaminduR**
### Proposed changes in this pull request

Add `getFederatedAuthSessionDetails` for retrieving federated session info during the federated idp initiated logout

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 1]
```diff
@@ -120,6 +120,10 @@
     public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING "
             + "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES (?, ?, ?, ?, ?)";
 
+    // Get federated authentication session details using the IDP session id
+    public static final String SQL_GET_FEDERATED_AUTH_SESSION_INFO_BY_SESSION_ID =
+            "SELECT * FROM IDN_FED_AUTH_SESSION_MAPPING WHERE IDP_SESSION_ID = ?";
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java:125]: Use the column names in the select

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 2]
```diff
@@ -89,4 +90,34 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
...
+    public FederatedUserSession getFederatedAuthSessionDetails(String fedIdpSessionId)
+            throws SessionManagementServerException {
+
+        FederatedUserSession federatedUserSession;
+        JdbcTemplate jdbcTemplate = JdbcUtils.getNewTemplate();
+        try {
+            federatedUserSession = jdbcTemplate
+                    .fetchSingleRecord(SQLQueries.SQL_GET_FEDERATED_AUTH_SESSION_INFO_BY_SESSION_ID,
+                            (resultSet, rowNumber) -> new FederatedUserSession(
+                                    resultSet.getString(1),
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:110]: Shall we get the columns by name here to improve readability

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java]: Changed in 3ae4174

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java (hunk 1 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java]: Changed in 3ae4174

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtConstants.java [hunk 3]
```diff
@@ -28,6 +28,13 @@
     public static final String IP_ADDRESS = "IP";
     public static final String LOGIN_TIME = "Login Time";
 
+    //Federated authentication session details column names
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtConstants.java:31]:
```suggestion
    // Federated authentication session details column names.
```

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtConstants.java]: 95439f5

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.model;
+
+/**
+ * Represents a session of a federated user
+ */
+public class FederatedUserSession {
+
+    private String idPSessionId;
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java:26]:
```suggestion
    private String idpSessionId;
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    private String sessionId;
+    private String idPName;
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java:28]:
```suggestion
    private String idpName;
```

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java]: Changed in 087a6c6

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java (hunk 4 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java]: Changed in 087a6c6

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 6]
```diff
@@ -120,6 +120,11 @@
     public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING "
             + "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES (?, ?, ?, ?, ?)";
 
+    // Get federated authentication session details using the IDP session id
+    public static final String SQL_GET_FEDERATED_AUTH_SESSION_INFO_BY_SESSION_ID =
+            "SELECT IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE FROM " +
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java:125]:
```suggestion
            "SELECT IDP_SESSION_ID, SESSION_ID, IDP_NAME, AUTHENTICATOR_ID, PROTOCOL_TYPE FROM " +
```

**Review by mefarazath — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 7]
```diff
@@ -29,4 +30,12 @@
...
     UserSession getSession(String sessionId) throws SessionManagementServerException;
 
+    /**
+     * Get federated authentication session mapping info during the federated idp init logout
+     *
+     * @param fedIdpSessionId
+     * @return A map containing federated authentication session details
+     * @throws SessionManagementServerException
+     */
+    FederatedUserSession getFederatedAuthSessionDetails(String fedIdpSessionId) throws SessionManagementServerException;
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:40]: have the accessmodifier as default.
`default FederatedUserSession getFederatedAuthSessionDetails(..)` . If we don't have this modifier, if someone extended this interface, they might get issues when adding new public API in an interface

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 8]
```diff
@@ -89,4 +90,34 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
...
+                                    resultSet.getString(SessionMgtConstants.FEDERATED_IDP_SESSION_ID),
+                                    resultSet.getString(SessionMgtConstants.FEDERATED_SESSION_ID),
+                                    resultSet.getString(SessionMgtConstants.FEDERATED_IDP_NAME),
+                                    resultSet.getString(SessionMgtConstants.FEDERATED_AUTHENTICATOR_ID),
+                                    resultSet.getString(SessionMgtConstants.FEDERATED_PROTOCOL_TYPE)),
+                            preparedStatement -> preparedStatement.setString(1, fedIdpSessionId));
+            return federatedUserSession;
+        } catch (DataAccessException e) {
+            throw new SessionManagementServerException(
+                    SessionMgtConstants.ErrorMessages.ERROR_CODE_UNABLE_TO_GET_SESSION,
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:119]: Shall we add a new error message and error code for getting federated user session? 
ERROR_CODE_UNABLE_TO_GET__FED_USER_SESSION

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
+package org.wso2.carbon.identity.application.authentication.framework.dao.impl;
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:1]: Add license header

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+        setupSessionStore();
+        DataSource dataSource = mock(DataSource.class);
+        mockStatic(IdentityDatabaseUtil.class);
+        when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSource);
+        when(dataSource.getConnection()).thenReturn(getConnection(DB_NAME));
+        FederatedUserSession federatedUserSession = userSessionDAO.getFederatedAuthSessionDetails(IDP_SESSION_INDEX);
+        assertEquals(federatedUserSession.getSessionId(), SESSION_CONTEXT_KEY);
+    }
+}
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:132]: add new line

Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:132]:
```suggestion
}

```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    @Test
+    public void testGetFederatedAuthSessionDetails() throws Exception {
+
+        setupSessionStore();
+        DataSource dataSource = mock(DataSource.class);
+        mockStatic(IdentityDatabaseUtil.class);
+        when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSource);
+        when(dataSource.getConnection()).thenReturn(getConnection(DB_NAME));
+        FederatedUserSession federatedUserSession = userSessionDAO.getFederatedAuthSessionDetails(IDP_SESSION_INDEX);
+        assertEquals(federatedUserSession.getSessionId(), SESSION_CONTEXT_KEY);
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:130]: assert all the columns here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java (hunk 7 above)
Comment (piraveena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:40]: It exceeds 120 line width

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    private static Map<String, BasicDataSource> dataSourceMap = new HashMap<>();
+    private static final String DB_NAME = "testOIDCSLO";
+    private static String SESSION_CONTEXT_KEY = "02278824dfe9862d265e389365c0a71c365401672491b78c6ee7dd6fc44d8af4";
+    private static String IDP_SESSION_INDEX = "15043ffc-877d-4205-af41-9b107f7da38c";
+
+    @BeforeMethod
+    public void init() {
+
+        userSessionDAO = new UserSessionDAOImpl();
+
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:46]: remove extra new line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+            PreparedStatement statement2 = connection1.prepareStatement(query);
+            statement2.setString(1, "15043ffc-877d-4205-af41-9b107f7da38c");
+            ResultSet resultSet = statement2.executeQuery();
+            String result = null;
+            if (resultSet.next()) {
+                result = resultSet.getString("SESSION_ID");
+            }
+            assertEquals(SESSION_CONTEXT_KEY, result, "Failed to handle for valid input");
+        }
+
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:118]: remove extra new line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 14]
```diff
@@ -29,4 +30,12 @@
 
     UserSession getSession(String sessionId) throws SessionManagementServerException;
 
+    /**
+     * Get federated authentication session mapping info during the federated idp init logout
+     *
+     * @param fedIdpSessionId
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:36]: add description to the parameter fedIdpSessionId

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 15]
```diff
@@ -29,4 +30,12 @@
...
+     * @return A map containing federated authentication session details
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:37]:
```suggestion
     * @return A map containing federated authentication session details.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 16]
```diff
@@ -29,4 +30,12 @@
...
+     * @throws SessionManagementServerException
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:38]:
```suggestion
     * @throws SessionManagementServerException.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 17]
```diff
@@ -89,4 +90,34 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
         }
         return null;
     }
+
+    /**
+     * Get federated authentication session mapping info during the federated idp init logout
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:95]:
```suggestion
     * Get federated user session details mapped for federated IDP sessionId.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 18]
```diff
@@ -89,4 +90,34 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
...
+     *
+     * @param fedIdpSessionId
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:97]: add parameter description

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java [hunk 19]
```diff
@@ -89,4 +90,34 @@ public UserSession getSession(String sessionId) throws SessionManagementServerEx
...
+     * @return A FederatedUserSession containing federated authentication session details
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImpl.java:98]:
```suggestion
     * @return A FederatedUserSession containing federated authentication session details.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.model;
+
+/**
+ * Represents a session of a federated user
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java:22]:
```suggestion
 * Represents a session of a federated user.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 21]
```diff
@@ -120,6 +120,11 @@
     public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO = "INSERT INTO IDN_FED_AUTH_SESSION_MAPPING "
             + "(IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES (?, ?, ?, ?, ?)";
 
+    // Get federated authentication session details using the IDP session id
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java:123]:
```suggestion
    // Get federated authentication session details using the IDP session id.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtConstants.java [hunk 22]
```diff
@@ -28,6 +28,13 @@
     public static final String IP_ADDRESS = "IP";
     public static final String LOGIN_TIME = "Login Time";
 
+    // Federated authentication session details column names
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/SessionMgtConstants.java:31]:
```suggestion
    // Federated authentication session details column names.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java [hunk 23]
```diff
@@ -29,4 +30,12 @@
 
     UserSession getSession(String sessionId) throws SessionManagementServerException;
 
+    /**
+     * Get federated authentication session mapping info during the federated idp init logout
```
Comment (piraveena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/dao/UserSessionDAO.java:34]:
```suggestion
     * Get federated authentication session mapping info during the federated IDP init logout.
```

**Review by piraveena — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java [hunk 24]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (lashinijay) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/FederatedUserSession.java:2]: Shall we update the license header?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.sql.DataSource;
+
+import static org.mockito.Mockito.mock;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.api.mockito.PowerMockito.when;
+import static org.testng.Assert.*;
```
Comment (lashinijay) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:46]: Shall we avoid * imports and only import the required?

Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/test/java/org/wso2/carbon/identity/application/authentication/framework/dao/impl/UserSessionDAOImplTest.java:46]: Fixed in master branch

**Pull request merged**